yarn-project/end-to-end/src/e2e_amm.test.ts

```diff
@@ -0,0 +1,338 @@
+import { type AccountWallet, type DebugLogger, Fr, type Wallet } from '@aztec/aztec.js';
+import { AMMContract, type TokenContract } from '@aztec/noir-contracts.js';
+
+import { jest } from '@jest/globals';
+
+import { deployToken, mintTokensToPrivate } from './fixtures/token_utils.js';
+import { setup } from './fixtures/utils.js';
+
+const TIMEOUT = 120_000;
+
+describe('AMM', () => {
+  jest.setTimeout(TIMEOUT);
+
+  let teardown: () => Promise<void>;
+
+  let logger: DebugLogger;
+
+  let adminWallet: AccountWallet;
+  let liquidityProvider: AccountWallet;
+  let otherLiquidityProvider: AccountWallet;
+  let swapper: AccountWallet;
+
+  let token0: TokenContract;
+  let token1: TokenContract;
+  let liquidityToken: TokenContract;
+
+  let amm: AMMContract;
+
+  const INITIAL_AMM_TOTAL_SUPPLY = 100000n;
+
+  // We need a large token amount so that the swap fee (0.3%) is observable.
+  const INITIAL_TOKEN_BALANCE = 1_000_000_000n;
+
+  beforeAll(async () => {
+    ({
+      teardown,
+      wallets: [adminWallet, liquidityProvider, otherLiquidityProvider, swapper],
+      logger,
+    } = await setup(4));
+
+    token0 = await deployToken(adminWallet, 0n, logger);
+    token1 = await deployToken(adminWallet, 0n, logger);
+    liquidityToken = await deployToken(adminWallet, 0n, logger);
+
+    amm = await AMMContract.deploy(adminWallet, token0.address, token1.address, liquidityToken.address)
+      .send()
+      .deployed();
+
+    // TODO(#9480): consider deploying the token by some factory when the AMM is deployed, and making the AMM be the
+    // minter there.
+    await liquidityToken.methods.set_minter(amm.address, true).send().wait();
+
+    // We mint the tokens to both liquidity providers and the swapper
+    await mintTokensToPrivate(token0, adminWallet, liquidityProvider.getAddress(), INITIAL_TOKEN_BALANCE);
+    await mintTokensToPrivate(token1, adminWallet, liquidityProvider.getAddress(), INITIAL_TOKEN_BALANCE);
+
+    await mintTokensToPrivate(token0, adminWallet, otherLiquidityProvider.getAddress(), INITIAL_TOKEN_BALANCE);
+    await mintTokensToPrivate(token1, adminWallet, otherLiquidityProvider.getAddress(), INITIAL_TOKEN_BALANCE);
+
+    // Note that the swapper only holds token0, not token1
+    await mintTokensToPrivate(token0, adminWallet, swapper.getAddress(), INITIAL_TOKEN_BALANCE);
+  });
+
+  afterAll(() => teardown());
+
+  describe('full flow', () => {
+    // This is an integration test in which we perform an entire run of the happy path. Thorough unit testing is not
+    // included.
+
+    type Balance = {
+      token0: bigint;
+      token1: bigint;
+    };
+
+    async function getAmmBalances(): Promise<Balance> {
```

Comment:
This is reasonably fast now, but consider using `BatchCall` since it adds up with increasing number of tests